File: mq/src/share/cclient/client/XASession.cpp
/*
 * Copyright (c) 2000, 2017 Oracle and/or its affiliates. All rights reserved.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0, which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the
 * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
 * version 2 with the GNU Classpath Exception, which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 */

/*
 * @(#)XASession.cpp	1.2 10/23/07
 */ 

#include "XASession.hpp"
#include "Connection.hpp"

/*
 *
 */
XASession::XASession(Connection * const connectionArg,
                     const ReceiveMode  receiveModeArg,
                     MQMessageListenerBAFunc beforeMessageListenerArg, 
                     MQMessageListenerBAFunc afterMessageListenerArg, 
                     void * callbackDataArg
                     ) : Session(connectionArg, PR_FALSE, AUTO_ACKNOWLEDGE, receiveModeArg) 
{
  CHECK_OBJECT_VALIDITY();

  ASSERT (connectionArg->getIsXA() == PR_TRUE ); 
  this->isXA = PR_TRUE; 
  this->ackMode = SESSION_TRANSACTED; 
  this->transactionID = LL_Zero();
  this->xidIndex = mq_getXidIndex();
  this->beforeMessageListener = NULL;
  this->afterMessageListener = NULL;
  this->baMLCallbackData = NULL;
  if (this->receiveMode == SESSION_ASYNC_RECEIVE) {
      this->beforeMessageListener = beforeMessageListenerArg;
      this->afterMessageListener = afterMessageListenerArg;
      this->baMLCallbackData = callbackDataArg;
  }
  LOG_INFO(( CODELOC, XA_SWITCH_LOG_MASK, connectionArg->id(), MQ_SUCCESS,
             "XASession (0x%p) created.", this ));
  return;
}


MQMessageListenerBAFunc
XASession::getBeforeMessageListenerFunc()
{
  CHECK_OBJECT_VALIDITY();
  return this->beforeMessageListener;
  
}


MQMessageListenerBAFunc
XASession::getAfterMessageListenerFunc()
{
  CHECK_OBJECT_VALIDITY();
  return this->afterMessageListener;
}


void *
XASession::getMessageListenerBACallbackData()
{
  CHECK_OBJECT_VALIDITY();
  return this->baMLCallbackData;
}


MQError
XASession::writeJMSMessage(Message * const message, PRInt64 producerID)
{
  CHECK_OBJECT_VALIDITY();

  if (this->isClosed) return MQ_SESSION_CLOSED;

  if (this->isXA == PR_TRUE) {
    MQXID *xid = (MQXID *) PR_GetThreadPrivate(this->xidIndex);
    if (xid == NULL) { 
      return MQ_THREAD_OUTSIDE_XA_TRANSACTION;
    }
    if (LL_IS_ZERO(xid->transactionID) != 0) {
      return MQ_XA_SESSION_NO_TRANSATION;
    }
    this->transactionID = xid->transactionID;
    {
      Long transactionIDLong(this->transactionID);
      LOG_FINE(( CODELOC, XA_SWITCH_LOG_MASK, this->connection->id(), MQ_SUCCESS,
                "XASession::writeJMSMessage with transactionID=%s, in XASession (0x%p)",
                 transactionIDLong.toString(), this ));
    }
  }
  return Session::writeJMSMessage(message, producerID);
}

MQError
XASession::acknowledge(Message * message, PRBool fromMessageListener)
{
  CHECK_OBJECT_VALIDITY();

  if (this->isClosed)  return MQ_SESSION_CLOSED;

  if (this->isXA == PR_TRUE) {
    MQXID *xid = (MQXID *) PR_GetThreadPrivate(this->xidIndex);
    if (xid == NULL) {
      return MQ_THREAD_OUTSIDE_XA_TRANSACTION;
    }
    if (LL_IS_ZERO(xid->transactionID) != 0) {
      return MQ_XA_SESSION_NO_TRANSATION;
    }
    this->transactionID = xid->transactionID;
    {
      Long transactionIDLong(this->transactionID);
      LOG_FINE(( CODELOC, XA_SWITCH_LOG_MASK, this->connection->id(), MQ_SUCCESS,
                "XASession::acknowledge with transactionID=%s, in XASession (0x%p)",
                 transactionIDLong.toString(), this ));
    }
  }

  return Session::acknowledge(message, fromMessageListener);
}


/*
 *
 */
XASession::~XASession()
{
  CHECK_OBJECT_VALIDITY();
}

File: mq/main/javax-jms/src/main/java/javax/jms/XASession.java
/*
 * Copyright (c) 1997, 2017 Oracle and/or its affiliates. All rights reserved.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0, which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the
 * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
 * version 2 with the GNU Classpath Exception, which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 */

package javax.jms;

import javax.transaction.xa.XAResource;

/** The {@code XASession} interface extends the capability of 
  * {@code Session} by adding access to a JMS provider's support for the
  * Java Transaction API (JTA) (optional). This support takes the form of a 
  * {@code javax.transaction.xa.XAResource} object. The functionality of 
  * this object closely resembles that defined by the standard X/Open XA 
  * Resource interface.
  *
  * <P>An application server controls the transactional assignment of an 
  * {@code XASession} by obtaining its {@code XAResource}. It uses 
  * the {@code XAResource} to assign the session to a transaction, prepare 
  * and commit work on the transaction, and so on.
  *
  * <P>An {@code XAResource} provides some fairly sophisticated facilities 
  * for interleaving work on multiple transactions, recovering a list of 
  * transactions in progress, and so on. A JTA aware JMS provider must fully 
  * implement this functionality. This could be done by using the services 
  * of a database that supports XA, or a JMS provider may choose to implement 
  * this functionality from scratch.
  *
  * <P>A client of the application server is given what it thinks is a 
  * regular JMS {@code Session}. Behind the scenes, the application server 
  * controls the transaction management of the underlying 
  * {@code XASession}.
  *
  * <P>The {@code XASession} interface is optional.  JMS providers 
  * are not required to support this interface. This interface is for 
  * use by JMS providers to support transactional environments. 
  * Client programs are strongly encouraged to use the transactional support
  * available in their environment, rather than use these XA
  * interfaces directly. 
  *
  * @see javax.jms.Session
  * 
  * @version JMS 2.0
  * @since JMS 1.0
  * 
  */
 
public interface XASession extends Session {

   /** Gets the session associated with this {@code XASession}.
      *  
      * @return the  session object
      *  
      * @exception JMSException if an internal error occurs.
      *
      * @since JMS 1.1
      */ 
     Session
     getSession() throws JMSException;
  
    /** Returns an XA resource to the caller.
      *
      * @return an XA resource to the caller
      */

     XAResource
     getXAResource();

    /** Indicates whether the session is in transacted mode.
      *  
      * @return true
      *  
      * @exception JMSException if the JMS provider fails to return the 
      *                         transaction mode due to some internal error.
      */ 

    boolean
    getTransacted() throws JMSException;


    /** Throws a {@code TransactionInProgressException}, since it should 
      * not be called for an {@code XASession} object.
      *
      * @exception TransactionInProgressException if the method is called on 
      *                         an {@code XASession}.
      *                                     
      */

    void
    commit() throws JMSException;


    /** Throws a {@code TransactionInProgressException}, since it should 
      * not be called for an {@code XASession} object.
      *
      * @exception TransactionInProgressException if the method is called on 
      *                         an {@code XASession}.
      *                                     
      */

    void
    rollback() throws JMSException;
}
